src/components/swap/PendingModalContent/ProgressIndicator.tsx

```diff
@@ -0,0 +1,189 @@
+  margin: 4px 11px;
+`
+type ProgressIndicatorStep = Extract<
+  ConfirmModalState,
+  | ConfirmModalState.APPROVING_TOKEN
+  | ConfirmModalState.PERMITTING
+  | ConfirmModalState.PENDING_CONFIRMATION
+  | ConfirmModalState.WRAPPING
+  | ConfirmModalState.RESETTING_TOKEN_ALLOWANCE
+>
+export function ProgressIndicator({
+  steps,
+  currentStep,
+  trade,
+  swapResult,
+  wrapTxHash,
+  tokenApprovalPending = false,
+  revocationPending = false,
+  swapError,
+  inputTokenColor,
+  onRetryUniswapXSignature,
+}: {
+  steps: ProgressIndicatorStep[]
+  currentStep: ProgressIndicatorStep
+  trade?: InterfaceTrade
+  swapResult?: SwapResult
+  wrapTxHash?: string
+  tokenApprovalPending?: boolean
+  revocationPending?: boolean
+  swapError?: Error | string
+  inputTokenColor?: string
+  onRetryUniswapXSignature?: () => void
+}) {
+  const { chainId } = useWeb3React()
+  const nativeCurrency = useNativeCurrency(chainId)
+
+  const swapStatus = useSwapTransactionStatus(swapResult)
+  const order = useOrder(swapResult?.type === TradeFillType.UniswapX ? swapResult.response.orderHash : '')
+
+  const swapConfirmed = swapStatus === TransactionStatus.Confirmed || order?.status === UniswapXOrderStatus.FILLED
+  const wrapConfirmed = useIsTransactionConfirmed(wrapTxHash)
+
+  const swapPending = swapResult !== undefined && !swapConfirmed
+  const wrapPending = wrapTxHash != undefined && !wrapConfirmed
+  const transactionPending = revocationPending || tokenApprovalPending || wrapPending || swapPending
+
+  const [signatureExpiredErrorId, setSignatureExpiredErrorId] = useState('')
+  useEffect(() => {
+    if (swapError instanceof SignatureExpiredError && swapError.id !== signatureExpiredErrorId) {
+      setSignatureExpiredErrorId(swapError.id)
+      onRetryUniswapXSignature?.()
+    }
+  }, [onRetryUniswapXSignature, signatureExpiredErrorId, swapError])
+
+  function getStatus(targetStep: ProgressIndicatorStep) {
+    const currentIndex = steps.indexOf(currentStep)
+    const targetIndex = steps.indexOf(targetStep)
+    if (currentIndex < targetIndex) {
+      return StepStatus.PREVIEW
+    } else if (currentIndex === targetIndex) {
+      return transactionPending ? StepStatus.IN_PROGRESS : StepStatus.ACTIVE
+    } else {
+      return StepStatus.COMPLETE
+    }
+  }
+
+  // TODO: update timeToEnd with dynamic value, calculated as approx. the most recent block confirmation time
+  const stepDetails: Record<ProgressIndicatorStep, StepDetails> = useMemo(
+    () => ({
+      [ConfirmModalState.WRAPPING]: {
+        icon: <CurrencyLogo currency={trade?.inputAmount.currency} />,
+        rippleColor: inputTokenColor,
+        previewTitle: `Wrap ${nativeCurrency.symbol}`,
+        actionRequiredTitle: `Wrap  ${nativeCurrency.symbol} in wallet`,
+        inProgressTitle: `Wrapping  ${nativeCurrency.symbol}...`,
+        timeToEnd: chainId === ChainId.MAINNET ? 20 : 10,
+        delayedTitle: 'Longer than expected...',
```

Comment:
the delayedTitle text is always the same, does this need to be a passed in param or part of the `Step` component?

Reply:
I think the `Step` component should remain generic, so steps / workflows can be added and customized without making changes to it.